store/user/store.ts

```diff
@@ -63,34 +77,87 @@ export const useUserStore = defineStore('userStore', () => {
     return computed(() => progressionByUnitId(unitId).value?.goals || []);
   }
 
-  function createDefaultProgression(unitId: Unit['id']) {
-    const unit = unitStore.unitById(unitId);
-    return {
-      goals: [],
-      date_completed: '',
-      unitId,
-      totalNumOfGoals: unit.value?.goals.length || 0,
-    };
+  async function generateProgressionAndGoal(
+    unitId: Unit['id'],
+    goalFromWidget: Goal
+  ) {
+    let progression: Progression;
+    if (isLoggedIn()) {
+      // TODO progress.date_completed: is not being used right now
+      const progressionDTO = await createProgressionAndGoal(
+        unitId,
+        goalFromWidget.custom_id,
+        null,
+        goalFromWidget.date_completed,
+        goalFromWidget.additional_data
+      );
+      progression = progressionDTOToProgression(progressionDTO);
+    } else {
+      const unit = unitStore.unitById(unitId);
+      progression = {
+        goals: [goalFromWidget],
+        date_completed: null, // TODO progress.date_completed: is not being used right now
+        unitId,
+        totalNumOfGoals: unit.value?.goals?.length || 0,
+      };
+    }
+    return progression;
+  }
+
+  async function updateOrCreateGoal(
+    progression: Progression,
+    goalFromWidget: Goal
+  ) {
+    if (!isLoggedIn()) {
+      return goalFromWidget;
+    }
+
+    const existingGoal = progression.goals.find(
+      (goal) => goal.custom_id === goalFromWidget.custom_id
+    );
+
+    const goalDTO = await (existingGoal
+      ? updateGoal(
+          existingGoal.id,
+          goalFromWidget.attempts,
+          goalFromWidget.date_completed,
+          null
+        )
+      : createGoal(
+          progression.id,
+          goalFromWidget.custom_id,
+          goalFromWidget.date_completed,
+          null
+        ));
+
+    return goalDTOToGoal(goalDTO);
   }
 
-  function setUnitProgressionGoal(unitId: string, goal: Goal) {
+  async function setUnitProgressionGoalFromWidget(
+    unitId: string,
+    widgetGoal: WidgetGoal
+  ) {
     let progression = progressionByUnitId(unitId).value;
+    let goal: Goal;
 
     if (!progression) {
-      progression = createDefaultProgression(unitId);
+      progression = await generateProgressionAndGoal(unitId, widgetGoal);
       setUnitProgression(progression);
+      goal = progression.goals[0];
```

Comment:
Could goals ever be undefined? 

```suggestion
      goal = progression.goals?.[0];
```

Reply:
I would say not in this case because you are creating both entities at the same time, `progression` and `goal`. Must be always one and only one.